src/mlpack/methods/cf/cf_impl.hpp

```diff
@@ -148,22 +130,271 @@ void CF::Train(const arma::sp_mat& data,
+    arma::Mat<size_t>& recommendations,
+    const arma::Col<size_t>& users)
+{
+  // We want to avoid calculating the full rating matrix, so we will do nearest
+  // neighbor search only on the H matrix, using the observation that if the
+  // rating matrix X = W*H, then d(X.col(i), X.col(j)) = d(W H.col(i), W
+  // H.col(j)).  This can be seen as nearest neighbor search on the H matrix
+  // with the Mahalanobis distance where M^{-1} = W^T W.  So, we'll decompose
+  // M^{-1} = L L^T (the Cholesky decomposition), and then multiply H by L^T.
+  // Then we can perform nearest neighbor search.
+  arma::mat l = arma::chol(w.t() * w);
+  arma::mat stretchedH = l * h; // Due to the Armadillo API, l is L^T.
+
+  // Now, we will use the decomposed w and h matrices to estimate what the user
+  // would have rated items as, and then pick the best items.
+
+  // Temporarily store feature vector of queried users.
+  arma::mat query(stretchedH.n_rows, users.n_elem);
+
+  // Select feature vectors of queried users.
+  for (size_t i = 0; i < users.n_elem; i++)
+    query.col(i) = stretchedH.col(users(i));
+
+  // Temporary storage for neighborhood of the queried users.
+  arma::Mat<size_t> neighborhood;
+
+  // Calculate the neighborhood of the queried users.  Note that the query user
+  // is part of the neighborhood---this is intentional.  We want to use an
+  // average of both the query user and the local neighborhood of the query
+  // user.
+  // The neighbor search technique should be a template parameter.
+  neighbor::KNN a(stretchedH);
+  arma::mat resultingDistances; // Temporary storage.
+  a.Search(query, numUsersForSimilarity, neighborhood, resultingDistances);
+
+  // Generate recommendations for each query user by finding the maximum numRecs
+  // elements in the averages matrix.
+  recommendations.set_size(numRecs, users.n_elem);
+  arma::mat values(numRecs, users.n_elem);
+  recommendations.fill(SIZE_MAX);
+  values.fill(DBL_MAX);
+
+  for (size_t i = 0; i < users.n_elem; i++)
+  {
+    // First, calculate average of neighborhood values.
+    arma::vec averages;
+    averages.zeros(cleanedData.n_rows);
+
+    for (size_t j = 0; j < neighborhood.n_rows; ++j)
+      averages += w * h.col(neighborhood(j, i));
+    averages /= neighborhood.n_rows;
+
+    // Let's build the list of candidate recomendations for the given user.
+    // Default candidate: the smallest possible value and invalid item number.
+    const Candidate def = std::make_pair(-DBL_MAX, cleanedData.n_rows);
+    std::vector<Candidate> vect(numRecs, def);
+    typedef std::priority_queue<Candidate, std::vector<Candidate>, CandidateCmp>
+        CandidateList;
+    CandidateList pqueue(CandidateCmp(), std::move(vect));
+
+    // Look through the averages column corresponding to the current user.
+    for (size_t j = 0; j < averages.n_rows; ++j)
+    {
+      // Ensure that the user hasn't already rated the item.
+      if (cleanedData(j, users(i)) != 0.0)
```

Comment:
After data normalization, an existing rating could be normalized to zero. (e.g. using OverallMeanNormalization and `r_ij == mean == 3`). Is it a good idea to use another sparse matrix (`arma::sp_umat`) to  save whether a rating is already given in training set?

Reply:
Hmm, that's an interesting question. I think you could denormalize the item here, however I am not sure it's the most cost-efficient way.

Reply:
@lozhnikov Whether the item is in training set doesn't change the way it is denormalized. So I don't think it works.

Reply:
yeah, you're right. There is another option: we could replace every exact zero by `std::numerical_limits<double>::min()` at the normalization step. However this solution looks like a crutch.

Reply:
That should be an efficient way to solve this. Let's use this solution for now and we can switch to the first solution if we run into any issue.